share/man/man4/agr.4
.\"	$NetBSD$
.\"
.\" Copyright (c)2005, 2007 YAMAMOTO Takashi,
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" ------------------------------------------------------------
.Dd February 23, 2010
.Dt AGR 4
.Os
.\" ------------------------------------------------------------
.Sh NAME
.Nm agr
.Nd link aggregation pseudo network interface driver
.\" ------------------------------------------------------------
.Sh SYNOPSIS
.Cd pseudo-device agr
.\" ------------------------------------------------------------
.Sh DESCRIPTION
The
.Nm
driver provides link aggregation functionality (a.k.a. L2 trunking
or bonding).
.Pp
It supports the IEEE 802.3ad Link Aggregation Control Protocol
(LACP) and the Marker Protocol.
.Pp
The
.Nm
driver supports the following link specific flags for
.Xr ifconfig 8 :
.Bl -tag -width "-link0"
.It Cm link0
Use the round-robin distribution algorithm.
Don't use it unless you're really sure,
because it violates the frame ordering rule.
.It Cm -link0
Use the default distribution algorithm, which is based on the hash
of DA/SA, TCI, and, if available, some upper layer protocol
information like
.Xr ip 4
DA/SA.
.It Cm link1
Disable LACP.
Prevents any LACP or Marker messaging which leaves
the ports in the default static configuration.
Set this prior to adding ports.
.El
.\" ------------------------------------------------------------
.Sh EXAMPLES
Create an
.Nm
interface,
.Sy agr0 ,
and attach
.Sy re0
and
.Sy re1
to it.
In other words, aggregate
.Sy re0
and
.Sy re1
so that they can be used as a single interface,
.Sy agr0 .
The physical interfaces which are attached to the
.Nm
interface must not have any IP addresses, neither IPv4 nor IPv6.
.Bd -literal -offset indent
	ifconfig re0 inet xxx.xxx.xxx.xxx delete
	ifconfig re0 inet6 fe80::xxxx:xxxx:xxxx:xxxx delete
	ifconfig re1 inet xxx.xxx.xxx.xxx delete
	ifconfig re1 inet6 fe80::xxxx:xxxx:xxxx:xxxx delete

	ifconfig agr0 create
	ifconfig agr0 agrport re0
	ifconfig agr0 agrport re1
.Ed
.Pp
Destroy an interface created in the above example.
.Bd -literal -offset indent
	ifconfig agr0 -agrport re0
	ifconfig agr0 -agrport re1
	ifconfig agr0 destroy
.Ed
.\" ------------------------------------------------------------
.Sh SEE ALSO
.Xr ifconfig 8
.\" ------------------------------------------------------------
.Sh STANDARDS
IEEE 802.3ad Aggregation of Multiple Link Segments
.\" ------------------------------------------------------------
.Sh HISTORY
The
.Nm
driver first appeared in
.Nx 4.0 .
.\" ------------------------------------------------------------
.Sh AUTHORS
The
.Nm
driver was written by
.An YAMAMOTO Takashi .
.\" ------------------------------------------------------------
.Sh BUGS
There is no way to configure LACP administrative variables, including
system and port priorities.
The current implementation of the
.Nm
driver always performs active-mode LACP and
uses 0x8000 as system and port priorities.
.Pp
The
.Nm
driver uses the MAC address of the first-added physical
interface as the MAC address of the
.Nm
interface itself.
Thus, removing the physical interface and using it for another purpose
can result in non-unique MAC addresses.
.Pp
The current implementation of the
.Nm
driver doesn't prevent unsafe operations like some ioctls against
underlying physical interfaces.
Such operations can result in unexpected behaviors, and are strongly
discouraged.
.Pp
There is no way to configure
.Nm
interfaces without attaching physical interfaces.
.Pp
Physical interfaces being added to the
.Nm
interface shouldn't have any addresses except for link level address.
Otherwise, the attempt will fail with
.Dv EBUSY .
Note that it includes an automatically assigned IPv6 link-local address.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]
	bge(4): Add support for BCM57762, found in Apple's Thunderbolt to
		Gigabit Ethernet Adapter. PR/46961 [tsutsui 20120917]
	mfi(4): add a command pass-through ioctl, and associated COMPAT_LINUX
		support. This allows to use the LSI MegaCLI linux binary
		on a NetBSD host.  [bouyer 20120919]
	zoneinfo: Import tzdata2012f. [apb 20120927]
	kernel: add clock_nanosleep(2) [christos 20121001]
	wpa:	New wpa_supplicant(8) and hostapd(8) 1.0 [christos 20121007]

sbin/ifconfig/agr.c
/*	$NetBSD$	*/

/*-
 * Copyright (c)2005 YAMAMOTO Takashi,
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if !defined(lint)
__RCSID("$NetBSD$");
#endif /* !defined(lint) */

#include <sys/param.h>
#include <sys/ioctl.h>

#include <net/if.h>
#include <net/agr/if_agrioctl.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <util.h>

#include "env.h"
#include "extern.h"
#include "parse.h"
#include "util.h"

static int agrsetport(prop_dictionary_t, prop_dictionary_t);
static void agr_constructor(void) __attribute__((constructor));
static int checkifname(prop_dictionary_t);
static void assertifname(prop_dictionary_t);

static struct piface agrif = PIFACE_INITIALIZER(&agrif, "agr interface",
    agrsetport, "agrport", &command_root.pb_parser);

static const struct kwinst agrkw[] = {
	  {.k_word = "agrport", .k_type = KW_T_INT, .k_int = AGRCMD_ADDPORT,
	   .k_nextparser = &agrif.pif_parser}
	, {.k_word = "-agrport", .k_type = KW_T_INT, .k_int = AGRCMD_REMPORT,
	   .k_nextparser = &agrif.pif_parser}
};

struct pkw agr = PKW_INITIALIZER(&agr, "agr", NULL, "agrcmd",
    agrkw, __arraycount(agrkw), NULL);

static int
checkifname(prop_dictionary_t env)
{
	const char *ifname;

	if ((ifname = getifname(env)) == NULL)
		return 1;

	return strncmp(ifname, "agr", 3) != 0 ||
	    !isdigit((unsigned char)ifname[3]);
}

static void
assertifname(prop_dictionary_t env)
{
	if (checkifname(env))
		errx(EXIT_FAILURE, "valid only with agr(4) interfaces");
}

int
agrsetport(prop_dictionary_t env, prop_dictionary_t oenv)
{
	char buf[IFNAMSIZ];
	struct agrreq ar;
	const char *port;
	int64_t cmd;

	if (!prop_dictionary_get_int64(env, "agrcmd", &cmd)) {
		warnx("%s.%d", __func__, __LINE__);
		errno = ENOENT;
		return -1;
	}

	if (!prop_dictionary_get_cstring_nocopy(env, "agrport", &port)) {
		warnx("%s.%d", __func__, __LINE__);
		errno = ENOENT;
		return -1;
	}
	strlcpy(buf, port, sizeof(buf));

	assertifname(env);
	memset(&ar, 0, sizeof(ar));
	ar.ar_version = AGRREQ_VERSION;
	ar.ar_cmd = cmd;
	ar.ar_buf = buf;
	ar.ar_buflen = strlen(buf);

	if (indirect_ioctl(env, SIOCSETAGR, &ar) == -1)
		err(EXIT_FAILURE, "SIOCSETAGR");
	return 0;
}

static void
agr_status(prop_dictionary_t env, prop_dictionary_t oenv)
{
	struct agrreq ar;
	void *buf = NULL;
	size_t buflen = 0;
	struct agrportlist *apl;
	struct agrportinfo *api;
	int i;

	if (checkifname(env))
		return;

again:
	memset(&ar, 0, sizeof(ar));
	ar.ar_version = AGRREQ_VERSION;
	ar.ar_cmd = AGRCMD_PORTLIST;
	ar.ar_buf = buf;
	ar.ar_buflen = buflen;

	if (indirect_ioctl(env, SIOCGETAGR, &ar) == -1) {
		if (errno != E2BIG) {
			warn("SIOCGETAGR");
			return;
		}

		free(buf);
		buf = NULL;
		buflen = 0;
		goto again;
	}

	if (buf == NULL) {
		buflen = ar.ar_buflen;
		buf = malloc(buflen);
		if (buf == NULL) {
			err(EXIT_FAILURE, "agr_status");
		}
		goto again;
	}

	apl = buf;
	api = (void *)(apl + 1);

	for (i = 0; i < apl->apl_nports; i++) {
		char tmp[256];

		snprintb(tmp, sizeof(tmp), AGRPORTINFO_BITS, api->api_flags);
		printf("\tagrport: %s, flags=%s\n", api->api_ifname, tmp);
		api++;
	}
}

static status_func_t status;
static usage_func_t usage;
static cmdloop_branch_t branch;

static void
agr_usage(prop_dictionary_t env)
{
	fprintf(stderr, "\t[ agrport i ] [ -agrport i ]\n");
}

static void
agr_constructor(void)
{
	status_func_init(&status, agr_status);
	usage_func_init(&usage, agr_usage);
	register_status(&status);
	register_usage(&usage);
	cmdloop_branch_init(&branch, &agr.pk_parser);
	register_cmdloop_branch(&branch);
}
